README.md
# varnishbackend_exporter

## Summary

`varnishbackend_exporter` is an exporter for `prometheus` that counts
backend in different states in Varnish. It does so by connecting to
the Varnish management interface using TCP (typically on
localhost:6082) and collects the statistics at regular intervals
(they are not collected on scrape, but are cached locally in order
to respond quickly to the scrapes).

## Exported metrics

`varnishbackend_exporter` exports a single metric, with multiple labels.
The metric name is `varnish_backend_state`. In the simplest mode, only
one label is attached, `state`. This can be either `healthy` or `sick`,
and contains the count of backends in this state at the given moment.

When run in `director regexp mode`, it will also export a label named
`director`, which will be set to the name captured using the regexp
(see below).


### director regexp mode

The Varnish administration interface does not directly expose which
backends are used in which directors. But if consistent naming is used,
then `varnishbackend_exporter` can extract the director name (or any
other data of course, but it will be labeled as director) from the name
of the backend.

This extraction is done using a regular expression given to the
`-directorre` parameter. This regular expression must have a capture
group, and the value captured in this group will be the one labeled
as the director name.

For example, if a naming standard has backend names like
`servername_systemname`, we can extract systemname by passing
`-directorre ".*_([^_]+)$"`.

Any backend not being matched by the regexp will be labeled as `unknown`.


## Usage

    -directorre string
      	Regular expression extracting director name from backend name
    -varnish.interval int
      	Varnish checking interval (default 15)
    -varnish.port int
      	Port of Varnish to connect to (default 6082)
    -varnish.secret string
      	Filename of varnish secret file (default "/etc/varnish/secret")
    -version
      	Print version information.
    -web.listen-address string
      	Address to listen on for web interface and telemetry. (default ":9133")
    -web.telemetry-path string
      	Path under which to expose metrics. (default "/metrics")


CHANGELOG.md
1.0
===
* Initial release
* Added a build script to build all for all platforms


build.sh
#!/bin/bash

set -e

if [ ! -e varnishbackend_exporter.go ] ; then
    echo "Error: Script can only be ran on the root of the source tree"
    exit 1
fi

rm -rf bin
mkdir -p bin/build bin/release

VERSION=$1
VERSION_HASH="$(git rev-parse --short HEAD)"
VERSION_DATE="$(date -u '+%d.%m.%Y %H:%M:%S')"

echo -e "\nVERSION=$VERSION"
echo "VERSION_HASH=$VERSION_HASH"
echo "VERSION_DATE=$VERSION_DATE"

if [ -z "$VERSION" ]; then
    echo "Error: First argument must be release version"
    exit 1
fi

#tar -cvzf ./bin/release/dashboards-$VERSION.tar.gz dashboards/* > /dev/null 2>&1

for goos in linux darwin windows freebsd openbsd netbsd ; do
    for goarch in amd64 386; do
        # path
        file_versioned="prometheus_varnishbackend_exporter-$VERSION.$goos-$goarch"
        outdir="bin/build/$file_versioned"
        path="$outdir/prometheus_varnishbackend_exporter"
        if [ $goos = windows ] ; then
            path=$path.exe
        fi

        mkdir -p $outdir
        cp LICENSE CHANGELOG.md README.md $outdir/

        # build
        echo -e "\nBuilding $goos/$goarch"
        GOOS=$goos GOARCH=$goarch go build -o $path -ldflags "-X 'main.Version=$VERSION' -X 'main.VersionHash=$VERSION_HASH' -X 'main.VersionDate=$VERSION_DATE'"
        echo "  > `du -hc $path | awk 'NR==1{print $1;}'`    `file $path`"

        # compress (for unique filenames to github release files)
        tar -C ./bin/build -cvzf ./bin/release/$file_versioned.tar.gz $file_versioned > /dev/null 2>&1
    done
done

go env > .goenv
source .goenv
rm .goenv

echo -e "\nRelease done: $(./bin/build/prometheus_varnishbackend_exporter-$VERSION.$GOOS-$GOARCH/prometheus_varnishbackend_exporter --version)"
for goos in linux darwin windows freebsd openbsd netbsd ; do
    for goarch in amd64 386; do
        file_versioned="prometheus_varnishbackend_exporter-$VERSION.$goos-$goarch"
        path=bin/release/$file_versioned.tar.gz
        echo "  > `du -hc $path | awk 'NR==1{print $1;}'`    $path"
    done
done

cd ./bin/release
sha256sum --binary ./* | sed -En "s/\*\.\/(.*)$/\1/p" > sha256sums.txt
cd ../..


varnishbackend_exporter.go
package main

import (
	"bufio"
	"crypto/sha256"
	"encoding/hex"
	"flag"
	"fmt"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/version"
	"io/ioutil"
	"net"
	"net/http"
	"os"
	"regexp"
	"strings"
	"time"
)

type VarnishWrapper struct {
	conn net.Conn
}

func (v *VarnishWrapper) ReadResponse() (code int, response *string) {
	var status, length int

	headers, err := fmt.Fscanf(v.conn, "%03d %8d\n", &status, &length)
	if err != nil {
		fmt.Printf("Failed to scan header: %s\n", err)
		return -1, nil
	}

	if headers != 2 {
		fmt.Printf("Invalid number of headers: %d\n", headers)
		return -1, nil
	}

	buf := make([]byte, length+1)
	l, err := v.conn.Read(buf)
	if err != nil {
		fmt.Printf("Read from Varnish failed: %s\n", err)
		return -1, nil
	}

	if l != length+1 {
		fmt.Printf("Read %d, expected %d\n", l, length+1)
		return -1, nil
	}

	ret := string(buf[0 : len(buf)-1])
	return status, &ret
}

func (v *VarnishWrapper) Send(str string, args ...string) error {
	var buf = append([]string{str}, args...)
	body := fmt.Sprintf("%s\n", strings.Join(buf, " "))
	_, err := v.conn.Write([]byte(body))
	if err != nil {
		fmt.Printf("Write error: %s\n", err)
		return err
	}
	return nil
}

func (v *VarnishWrapper) CommandForSuccess(cmd string, args ...string) bool {
	err := v.Send(cmd, args...)
	if err != nil {
		return false
	}
	code, _ := v.ReadResponse()
	return (code == 200)
}

/* Prometheus counters */
var prombackends *prometheus.GaugeVec

var directorRegexp *regexp.Regexp = nil
var promlabels []string

/* Webserver goroutine that servers up the current metrics */
func httpServer(listenAddress string, metricsPath string) {
	http.Handle(metricsPath, promhttp.Handler())
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`<html>
             <head><title>Varnishbackend Exporter</title></head>
             <body>
             <h1>Varnishbackend Exporter</h1>
             <p><a href='` + metricsPath + `'>Metrics</a></p>
             </body>
             </html>`))
	})
	http.ListenAndServe(listenAddress, nil)
}

var (
	debug = flag.Bool("debug", false, "Print debugging information.")
)

func Debug(msg string) {
	if *debug {
		fmt.Println(msg)
	}
}

func main() {
	var (
		listenAddress   = flag.String("web.listen-address", ":9133", "Address to listen on for web interface and telemetry.")
		metricsPath     = flag.String("web.telemetry-path", "/metrics", "Path under which to expose metrics.")
		varnishPort     = flag.Int("varnish.port", 6082, "Port of Varnish to connect to")
		varnishSecret   = flag.String("varnish.secret", "/etc/varnish/secret", "Filename of varnish secret file")
		varnishInterval = flag.Int("varnish.interval", 15, "Varnish checking interval")
		directorReStr   = flag.String("directorre", "", "Regular expression extracting director name from backend name")
		showVersion     = flag.Bool("version", false, "Print version information.")
	)
	flag.Parse()

	if *showVersion {
		fmt.Println(version.Print("varnishbackend_exporter"))
		os.Exit(0)
	}

	if *directorReStr != "" {
		directorRegexp = regexp.MustCompile(*directorReStr)
		promlabels = []string{"state", "director"}
	} else {
		promlabels = []string{"state"}
	}

	secret, err := ioutil.ReadFile(*varnishSecret)
	if err != nil {
		fmt.Printf("Failed to read %s: %s\n", *varnishSecret, err)
		os.Exit(1)
	}

	prombackends = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "varnish_backend_state",
			Help: "varnish backend states",
		},
		promlabels,
	)
	prometheus.MustRegister(prombackends)

	// Http listener
	go httpServer(*listenAddress, *metricsPath)

	// Main loop to poll Varnish
	tcpAddr, err := net.ResolveTCPAddr("tcp", fmt.Sprintf("localhost:%d", *varnishPort))
	if err != nil {
		fmt.Printf("Could not resolve address: %s\n", err)
		os.Exit(1)
	}

	first := true
	for {
		/* To make sure we don't flood things */
		if first {
			first = false
		} else {
			/* Rate limit */
			Debug("Sleeping 5 seconds before connecting")
			time.Sleep(5 * time.Second)
		}
		Debug("Connecting to Varnish")
		conn, err := net.DialTCP("tcp", nil, tcpAddr)
		if err != nil {
			fmt.Printf("Connection failed: %s\n", err.Error())
			continue
		}
		defer conn.Close()
		vadm := &VarnishWrapper{conn: conn}
		code, resp := vadm.ReadResponse()
		if code != 107 {
			fmt.Println("Varnish did not give authentication prompt.")
			continue
		}
		challenge := strings.Split(*resp, "\n")[0]
		response := sha256.Sum256([]byte(fmt.Sprintf("%s\n%s%s\n", challenge, secret, challenge)))
		if !vadm.CommandForSuccess("auth", hex.EncodeToString(response[:])) {
			fmt.Println("Failed to authenticate")
			continue
		}

		/*
		 * Now that we have a working connection, loop with the same
		 * connection for multiple commands.
		 */
		for {
			Debug("Getting list from Varnish")
			err := vadm.Send("backend.list")
			if err != nil {
				break
			}

			code, resp := vadm.ReadResponse()
			if code != 200 {
				fmt.Printf("Received code %d, expected 200\n", code)
				break
			}
			scanner := bufio.NewScanner(strings.NewReader(*resp))
			var healthy, sick int
			var labelhealthy, labelsick, labelall map[string]int
			if directorRegexp != nil {
				labelhealthy = make(map[string]int)
				labelsick = make(map[string]int)
				labelall = make(map[string]int)
			} else {
				healthy = 0
				sick = 0
			}

			for scanner.Scan() {
				t := scanner.Text()
				if strings.HasPrefix(t, "Backend name ") {
					continue
				}
				fields := strings.Fields(t)

				if directorRegexp != nil {
					var lbl string
					m := directorRegexp.FindStringSubmatch(fields[0])
					if m != nil && len(m) > 1 {
						lbl = m[1]
					} else {
						lbl = "unknown"
					}
					labelall[lbl] = 1
					if fields[1] != "sick" && fields[2] == "Healthy" {
						labelhealthy[lbl]++
					} else {
						labelsick[lbl]++
					}
				} else {
					if fields[1] != "sick" && fields[2] == "Healthy" {
						healthy++
					} else {
						sick++
					}
				}
			}
			if directorRegexp != nil {
				for k := range labelall {
					prombackends.With(prometheus.Labels{"state": "healthy", "director": k}).Set(float64(labelhealthy[k]))
					prombackends.With(prometheus.Labels{"state": "sick", "director": k}).Set(float64(labelsick[k]))
				}
			} else {
				prombackends.With(prometheus.Labels{"state": "healthy"}).Set(float64(healthy))
				prombackends.With(prometheus.Labels{"state": "sick"}).Set(float64(sick))
			}

			Debug(fmt.Sprintf("Sleeping for %d seconds.", *varnishInterval))
			time.Sleep(time.Duration(*varnishInterval) * time.Second)
		}

		conn.Close()
	}
}
